File: 98.validate-binary-search-tree.js
/*
 * @lc app=leetcode id=98 lang=javascript
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
// TODO: to finish
var isValidBST = function (root) {

};

var f = function (node) {
  // 左子树要给 最大值，右子树要给 最小值。
  // leftVal = 

  // 1. 判断是否符合 left < node && right > node
  /* if (node.left && node.left.val >= node.val) {
    return null; // 不是合法的二叉查找树。
  }
  if (node.right && node.right.val <= node.val) {
    return null; // 不是合法的二叉查找树。
  } */
  

  const leftData = f(node.left);
  const rightData = f(node.right);

  if (!leftData) min = 
  const min = leftData.min;
  const max = rightData.max;
  return {
    min, max
  }

};
// @lc code=end

File: 98.validate-binary-search-tree.c
/*
 * @lc app=leetcode id=98 lang=c
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#include<stdio.h>

bool found = false;

// TODO: to finish
bool isValidBST_(struct TreeNode* node, struct TreeNode* parent){
  // [2,1,3]
  if (found == 1) return 0;
  if (node == NULL) return 1;

  bool leftBool = isValidBST_(node->left, node);
  if (!leftBool) {
    found = 1;
    return 0;
  }
  bool rightBool = isValidBST_(node->right, node);
  if (!rightBool) {
    found = 1;
    return 0;
  }

  bool val = 1;
  if (parent->left == node && parent->val < node->val) { // node is left node
    val = 0;
  } else if (parent->right == node && parent->val > node->val) { // node is right node
    val = 0;
  } 
  return val;

  // return isValidBST(root->left) && isValidBST(root->right);
}

bool isValidBST(struct TreeNode* root){
  if (root == NULL) return 1;
  if (root->left && root->left->val >= root->val) return 0;
  if (root->right && root->right->val <= root->val) return 0;
  return isValidBST_(root->left, root) && isValidBST_(root->right, root);
}

// @lc code=end

